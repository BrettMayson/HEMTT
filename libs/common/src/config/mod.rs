#![allow(clippy::module_name_repetitions)]

//! Read the project configuration into a [`ProjectConfig`] struct

use std::path::Path;

use tracing::warn;

mod addon;
mod pdrive;
mod project;

pub use addon::AddonConfig;
pub use pdrive::PDriveOption;
pub use project::{hemtt::launch::LaunchOptions, ProjectConfig};

fn deprecated(file: &Path, key: &str, replacement: &str, info: Option<&str>) {
    warn!(
        "Use of deprecated key '{}' in '{}'. Use '{}' instead.{}",
        key,
        file.display(),
        replacement,
        info.map(|i| format!("\n  {}", i)).unwrap_or_default()
    );
}
