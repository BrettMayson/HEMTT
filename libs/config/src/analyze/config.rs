use std::collections::{HashMap, HashSet};

use hemtt_common::reporting::{Code, Processed};
use hemtt_project::ProjectConfig;

use crate::{Class, Config, Ident, Item, Property, Str, Value};

use super::{
    codes::{
        ce3_duplicate_property::DuplicateProperty, ce7_missing_parent::MissingParent,
        cw1_parent_case::ParentCase, cw2_magwell_missing_magazine::MagwellMissingMagazine,
    },
    Analyze,
};

impl Analyze for Config {
    fn valid(&self, project: Option<&ProjectConfig>) -> bool {
        self.0.iter().all(|p| p.valid(project))
    }

    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Box<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashMap::new();
        warnings.extend(external_parent_case_warn(&self.0, &mut defined));
        if let Some(project) = project {
            warnings.extend(magwell_missing_magazine(project, self));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Box<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashSet::new();
        errors.extend(external_missing_error(&self.0, &mut defined));
        errors.extend(duplicate_properties(&self.0));
        errors
    }
}

fn external_missing_error(
    properties: &[Property],
    defined: &mut HashSet<String>,
) -> Vec<Box<dyn Code>> {
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    if !defined.contains(&name) {
                        defined.insert(name);
                    }
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent = parent.value.to_lowercase();
                        if parent != name && !defined.contains(&parent) {
                            errors.push(Box::new(MissingParent::new(c.clone())));
                        }
                    }
                    defined.insert(name);
                    errors.extend(external_missing_error(properties, defined));
                }
            }
        }
    }
    errors
}

fn external_parent_case_warn(
    properties: &[Property],
    defined: &mut HashMap<String, Class>,
) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    defined.entry(name).or_insert_with(|| c.clone());
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name_lower = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent_lower = parent.value.to_lowercase();
                        if parent_lower != name_lower {
                            if let Some(parent_class) = defined.get(&parent_lower) {
                                if parent_class.name().value != parent.value {
                                    warnings.push(Box::new(ParentCase::new(
                                        c.clone(),
                                        parent_class.clone(),
                                    )));
                                }
                            }
                        } else if parent.value != name.value {
                            warnings.push(Box::new(ParentCase::new(c.clone(), c.clone())));
                        }
                    }
                    defined.insert(name_lower, c.clone());
                    warnings.extend(external_parent_case_warn(properties, defined));
                }
            }
        }
    }
    warnings
}

fn duplicate_properties(properties: &[Property]) -> Vec<Box<dyn Code>> {
    let mut seen: HashMap<String, Vec<(bool, Ident)>> = HashMap::new();
    duplicate_properties_inner("", properties, &mut seen);
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for (_, idents) in seen {
        if idents.len() > 1 && !idents.iter().all(|(class, _)| *class) {
            errors.push(Box::new(DuplicateProperty::new(
                idents.into_iter().map(|(_, i)| i).collect(),
            )));
        }
    }
    errors
}

fn duplicate_properties_inner(
    scope: &str,
    properties: &[Property],
    seen: &mut HashMap<String, Vec<(bool, Ident)>>,
) {
    for property in properties {
        match property {
            Property::Class(Class::Local {
                name, properties, ..
            }) => {
                duplicate_properties_inner(
                    &format!("{}.{}", scope, name.value.to_lowercase()),
                    properties,
                    seen,
                );
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((true, name.clone()));
            }
            Property::Entry { name, .. } | Property::MissingSemicolon(name, _) => {
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((false, name.clone()));
            }
            _ => (),
        }
    }
}

fn magwell_missing_magazine(project: &ProjectConfig, config: &Config) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    let mut classes = Vec::new();
    let Some(Property::Class(Class::Local {
        properties: magwells,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazinewells")
    else {
        return warnings;
    };
    let Some(Property::Class(Class::Local {
        properties: magazines,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazines")
    else {
        return warnings;
    };
    for property in magazines {
        if let Property::Class(Class::Local { name, .. }) = property {
            classes.push(name);
        }
    }
    for magwell in magwells {
        let Property::Class(Class::Local {
            properties: addons, ..
        }) = magwell
        else {
            continue;
        };
        for addon in addons {
            let Property::Entry {
                name,
                value: Value::Array(magazines),
                ..
            } = addon
            else {
                continue;
            };
            for mag in &magazines.items {
                let Item::Str(Str { value, span }) = mag else {
                    continue;
                };
                if !value
                    .to_lowercase()
                    .starts_with(&project.prefix().to_lowercase())
                {
                    continue;
                }
                if !classes.iter().any(|c| c.value == *value) {
                    warnings.push(Box::new(MagwellMissingMagazine::new(
                        name.clone(),
                        value.clone(),
                        span.clone(),
                    )));
                }
            }
        }
    }
    warnings
}
