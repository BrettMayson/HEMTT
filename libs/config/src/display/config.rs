use crate::Config;

impl std::fmt::Display for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for property in &self.0 {
            write!(f, "{property}")?;
        }
        Ok(())
    }
}
