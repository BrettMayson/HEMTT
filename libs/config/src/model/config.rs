use hemtt_common::version::Version;

use crate::{analyze::CfgPatch, Class, Number, Property, Value};

#[derive(Debug, PartialEq)]
/// A config file
pub struct Config(pub Vec<Property>);

impl Config {
    #[must_use]
    pub fn to_class(&self) -> Class {
        Class::Root {
            properties: self.0.clone(),
        }
    }
}

impl Config {
    #[must_use]
    /// Get the patches
    pub fn get_patches(&self) -> Vec<CfgPatch> {
        let mut patches = Vec::new();
        for property in &self.0 {
            if let Property::Class(Class::Local {
                name, properties, ..
            }) = property
            {
                if name.as_str() == "CfgPatches" {
                    for patch in properties {
                        if let Property::Class(Class::Local {
                            name, properties, ..
                        }) = patch
                        {
                            let mut required_version = Version::new(0, 0, 0, None);
                            for property in properties {
                                if let Property::Entry { name, value, .. } = property {
                                    if name.as_str() == "requiredVersion" {
                                        if let Value::Number(Number::Float32 { value, .. }) = value
                                        {
                                            required_version = Version::from(*value);
                                        }
                                    }
                                }
                            }
                            patches.push(CfgPatch::new(name.clone(), required_version));
                        }
                    }
                }
            }
        }
        patches
    }
}
