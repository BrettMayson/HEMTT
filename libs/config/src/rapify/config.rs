use std::io::Cursor;

use byteorder::{LittleEndian, WriteBytesExt};

use crate::{Class, Config};

use super::Rapify;

impl Rapify for Config {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_all(b"\0raP")?;
        output.write_all(b"\0\0\0\0\x08\0\0\0")?;

        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        let buffer: Box<[u8]> = vec![0; root_class.rapified_length()].into_boxed_slice();
        let mut cursor = Cursor::new(buffer);
        let written = root_class.rapify(&mut cursor, 16)?;
        assert_eq!(written, root_class.rapified_length());

        let enum_offset = 16 + cursor.get_ref().len() as u32;
        output.write_u32::<LittleEndian>(enum_offset)?;

        output.write_all(cursor.get_ref())?;

        output.write_all(b"\0\0\0\0")?;
        assert_eq!(written + 20, self.rapified_length());
        Ok(written + 20)
    }

    fn rapified_length(&self) -> usize {
        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        root_class.rapified_length() + 20 // metadata
    }
}
