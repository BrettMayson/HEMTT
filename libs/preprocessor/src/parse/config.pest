file = _{ SOI ~ (
    word | alpha | digit | underscore | dash | assignment | plus | left_brace | right_brace | left_bracket | right_bracket |
    left_parentheses | right_parentheses | colon | semicolon | join | directive | escape | slash | comma | decimal |
    double_quote | single_quote | left_angle | right_angle | unicode | newline
)* ~ EOI }
word = ${ ( underscore | alpha ) ~ ( alpha | digit | underscore )* }
alpha = { LETTER }
digit = { ASCII_DIGIT }
underscore = { "_" }
dash = { "-" }
assignment = { "=" }
plus = { "+" }
left_brace = { "{" }
right_brace = { "}" }
left_bracket = { "[" }
right_bracket = { "]" }
left_parentheses = { "(" }
right_parentheses = { ")" }
colon = { ":" }
semicolon = { ";" }
join = { "##" }
directive = { "#" }
escape = { "\\" }
slash = { "/" }
comma = { "," }
decimal = { "." }
double_quote = { "\"" }
single_quote = { "'" }
left_angle = { "<" }
right_angle = { ">" }
unicode = { LETTER | MARK | NUMBER | SYMBOL | PUNCTUATION | MATH | OTHER_MATH }
newline = { NEWLINE }

space = { " " }
tab = { "\t" }
WHITESPACE = ${ space | tab }
COMMENT = { ("/*" ~ (!"*/" ~ ANY)* ~ "*/") }
