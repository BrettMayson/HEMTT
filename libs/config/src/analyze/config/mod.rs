use std::sync::Arc;

use hemtt_common::config::ProjectConfig;
use hemtt_workspace::reporting::{Code, Processed};

use crate::Config;

use super::Analyze;

mod duplicate_classes;
mod duplicate_properties;
mod external_missing;
mod external_parent_case;
mod magwells;

impl Analyze for Config {
    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Arc<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        warnings.extend(external_parent_case::warn(&self.0, processed));
        if let Some(project) = project {
            warnings.extend(magwells::missing_magazine(project, self, processed));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Arc<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        errors.extend(duplicate_classes::error(&self.0, processed));
        errors.extend(external_missing::error(&self.0, processed));
        errors.extend(duplicate_properties::duplicate_properties(
            &self.0, processed,
        ));
        errors
    }
}
