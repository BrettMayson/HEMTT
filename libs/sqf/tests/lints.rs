#![allow(clippy::unwrap_used)]

use std::{io::Read, sync::Arc};

use hemtt_common::config::ProjectConfig;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{analyze::analyze, parser::database::Database};
use hemtt_workspace::{addons::Addon, reporting::WorkspaceFiles, LayerType};

const ROOT: &str = "tests/lints/";

macro_rules! analyze {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $dir>]() {
                test_analyze(stringify!($dir));
            }
        }
    };
}

fn test_analyze(dir: &str) {
    let folder = std::path::PathBuf::from(ROOT).join(dir);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join("source.sqf").unwrap();
    let processed = Processor::run(&source).unwrap();
    let database = Arc::new(Database::a3(false));
    let workspace_files = WorkspaceFiles::new();

    let config_path_full = std::path::PathBuf::from(ROOT).join("project_tests.toml");
    let config = ProjectConfig::from_file(&config_path_full).unwrap();

    match hemtt_sqf::parser::run(&database, &processed) {
        Ok(sqf) => {
            let codes = analyze(
                &sqf,
                Some(&config),
                &processed,
                Arc::new(Addon::test_addon()),
                database.clone(),
            );
            let stdout = codes
                .iter()
                .map(|e| e.diagnostic().unwrap().to_string(&workspace_files))
                .collect::<Vec<_>>()
                .join("\n")
                .replace('\r', "");
            let mut expected = Vec::new();
            std::fs::File::open(folder.join("stdout.ansi"))
                .unwrap()
                .read_to_end(&mut expected)
                .unwrap();
            if expected.is_empty() {
                std::fs::write(folder.join("stdout.ansi"), stdout.as_bytes()).unwrap();
            }
            let expected = String::from_utf8_lossy(&expected).replace('\r', "");
            assert_eq!(stdout, expected);
        }
        Err(hemtt_sqf::parser::ParserError::ParsingError(e)) => {
            for error in e {
                println!(
                    "{}",
                    error.diagnostic().unwrap().to_string(&workspace_files)
                );
            }
            panic!("failed to parse");
        }
        Err(e) => panic!("{e:?}"),
    };
}

analyze!(s02_event_handler_case);
analyze!(s03_static_typename);
analyze!(s04_command_case);
analyze!(s05_if_assign);
analyze!(s06_find_in_str);
analyze!(s07_select_parse_number);
analyze!(s08_format_args);
analyze!(s09_banned_command);
analyze!(s11_if_not_else);
analyze!(s17_var_all_caps);
analyze!(s18_in_vehicle_check);
analyze!(s20_bool_static_comparison);
analyze!(s21_invalid_comparisons);
analyze!(s22_this_call);
analyze!(s23_reassign_reserved_variable);
