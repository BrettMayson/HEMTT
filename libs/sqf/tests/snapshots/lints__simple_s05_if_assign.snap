---
source: libs/sqf/tests/lints.rs
expression: lint(stringify! (s05_if_assign))
---
[0m[1m[38;5;14mhelp[L-S05][0m[1m: assignment to if can be replaced with select[0m
  [0m[36mâ”Œâ”€[0m s05_if_assign.sqf:1:24
  [0m[36mâ”‚[0m
[0m[36m1[0m [0m[36mâ”‚[0m private _workingArms = [0m[36mif (alive player) then { 2 } else { 0 }[0m;
  [0m[36mâ”‚[0m                        [0m[36m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m [0m[36muse select[0m
  [0m[36mâ”‚[0m
  [0m[36m=[0m [36mnote[0m: the if and else blocks only return constant values
          select is faster in this case
  [0m[36m=[0m [32mtry[0m: [0, 2] select (alive player)


[0m[1m[38;5;14mhelp[L-S05][0m[1m: assignment to if can be replaced with select[0m
  [0m[36mâ”Œâ”€[0m s05_if_assign.sqf:4:22
  [0m[36mâ”‚[0m
[0m[36m4[0m [0m[36mâ”‚[0m     private _thing = [0m[36mif (alive player) then { 2 } else { 0 }[0m;
  [0m[36mâ”‚[0m                      [0m[36m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m [0m[36muse select[0m
  [0m[36mâ”‚[0m
  [0m[36m=[0m [36mnote[0m: the if and else blocks only return constant values
          select is faster in this case
  [0m[36m=[0m [32mtry[0m: [0, 2] select (alive player)


[0m[1m[38;5;14mhelp[L-S05][0m[1m: assignment to if can be replaced with select[0m
  [0m[36mâ”Œâ”€[0m s05_if_assign.sqf:7:5
  [0m[36mâ”‚[0m
[0m[36m7[0m [0m[36mâ”‚[0m     [0m[36mif (alive player) then { "torso" } else { "legs" }[0m,
  [0m[36mâ”‚[0m     [0m[36m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m [0m[36muse select[0m
  [0m[36mâ”‚[0m
  [0m[36m=[0m [36mnote[0m: the if and else blocks only return constant values
          select is faster in this case
  [0m[36m=[0m [32mtry[0m: ["legs", "torso"] select (alive player)


[0m[1m[38;5;14mhelp[L-S05][0m[1m: assignment to if can be replaced with select[0m
  [0m[36mâ”Œâ”€[0m s05_if_assign.sqf:8:5
  [0m[36mâ”‚[0m
[0m[36m8[0m [0m[36mâ”‚[0m     [0m[36mif (alive player) then { "torso" } else { "legs" }[0m
  [0m[36mâ”‚[0m     [0m[36m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m [0m[36muse select[0m
  [0m[36mâ”‚[0m
  [0m[36m=[0m [36mnote[0m: the if and else blocks only return constant values
          select is faster in this case
  [0m[36m=[0m [32mtry[0m: ["legs", "torso"] select (alive player)
